File: baseunits/modules/WPManga.pas
unit WPManga;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, synautil, synacode, RegExpr;

implementation

function GetDirURL(const AModule: TModuleContainer): String;
begin
  with AModule do
    if StringIn(Website, ['MangaSpy', 'MangaIce']) then
      Result := 'manga_list'
    else
    if Website = 'ReadHentaiManga' then
      Result := 'hentai-manga-list'
    else if Website = 'HentaiRead' then
      Result := 'hentai-list'
    else
      Result := 'manga-list';
  Result := '/' + Result + '/all/any/last-added/';
end;

function GetDirectoryPageNumber(const MangaInfo: TMangaInformation; var Page: Integer; const WorkPtr: Integer;
  const Module: TModuleContainer): Integer;
begin
  Page := 1;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL + GetDirURL(Module)) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        Page := StrToIntDef(ReplaceRegExpr('^.*\/(\d+)/.*$',
          XPathString('//ul[@class="pgg"]/li[last()]/a/@href'), '$1', True), 1);
      finally
        Free;
      end;
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation; const ANames, ALinks: TStringList;
  const AURL: String; const Module: TModuleContainer): Integer;
begin
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL + GetDirURL(Module) + IncStr(AURL) + '/') then begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        if StringIn(Module.Website, ['MangaSpy', 'MangaIce']) then
          XPathHREFAll('//*[contains(@id,"content")]//*[@class="det"]/a', ALinks, ANames)
        else
          XPathHREFtitleAll('//*[contains(@id,"content")]//a[./img]', ALinks, ANames);
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
  s: String;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := MaybeFillHost(Module.RootURL, AURL);
    if GET(url) then begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          coverLink := XPathString('//img[starts-with(@class,"cvr")]/@src');
          if coverLink <> '' then coverLink := MaybeFillHost(Module.RootURL, coverLink);
          title := XPathString('//*[@itemprop="itemreviewed"]');
          authors := XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Author")]/substring-after(.," ")');
          artists := XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Artist")]/substring-after(.," ")');
          status := MangaInfoStatusIfPos(XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Status")]/substring-after(.," ")'));
          if StringIn(Module.Website, ['ReadHentaiManga', 'HentaiRead']) then
            genres := XPathString('string-join(//*[contains(@class,"mng_det")]//*[self::p or self::li]//a,", ")')
          else
            genres := XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Category")]/string-join((./*[position()>1]),", ")');
          if Module.Website = 'HentaiRead' then
          begin
            chapterLinks.Add(XPathString('//a[@class="lst"]/@href'));
            chapterName.Add(title);
          end
          else
          begin
            while True do
            begin
              for v in XPath('//a[@class="lst"]') do
              begin
                chapterLinks.Add(v.toNode.getAttribute('href'));
                s := v.toNode.getAttribute('title');
                if s = '' then
                  s := XPathString('*[@class="val"]', v.toNode);
                if s = '' then
                  s := XPathString('text()[1]', v.toNode);
                chapterName.Add(s);
              end;
              if Thread.IsTerminated then Break;
              s := Trim(XPathString('//*[@class="pgg"]//*[./a[@class="sel"]]/following-sibling::*[./a]/a/@href'));
              if s = '' then Break;
              if GET(MaybeFillHost(Module.RootURL, s)) then
                ParseHTML(Document)
              else
                Break;
            end;
            InvertStrings([chapterLinks, chapterName]);
          end;
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread; const AURL: String;
  const Module: TModuleContainer): Boolean;
var
  s: String;
  v, x: IXQValue;
  allnum: Boolean;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.FHTTP, DownloadThread.Task.Container do begin
    PageLinks.Clear;
    PageNumber := 0;
    Cookies.Values['viewer'] := '1';
    if GET(AppendURLDelim(FillHost(Module.RootURL, AURL)) + '1') then begin
      Result := True;
      //multi page
      with TXQueryEngineHTML.Create(Document) do
        try
          s := XPathString('//script[contains(.,"imglist")]/substring-after(substring-before(.,"]"),"[")');
          if s <> '' then
            s := '[' + s + ']'
          else
          begin
            s := XPathString('//script[contains(.,"img_lst")]/substring-after(substring-before(.,"'')"),"(''")');
            if s <> '' then
              s := DecodeURL(s);
          end;
          ParseHTML(s);
          XPathStringAll('json(*)()("url")', PageLinks);
          if PageLinks.Count = 0 then
            XPathStringAll('json(*)()', PageLinks);
        finally
          Free;
        end;
      //single page
      if PageLinks.Count = 0 then
        with TXQueryEngineHTML.Create(Document) do
          try
            PageNumber := XPath('(//select[@class="cbo_wpm_pag"])[1]/option').Count;
            if PageNumber = 0 then
              PageNumber := XPath('(//select[@name="page"])[1]/option').Count;
            if PageNumber = 0 then
              for v in XPath('//select') do
              begin
                allnum := True;
                for x in XPath('option', v.toNode) do
                begin
                  if StrToIntDef(x.toString, -1) = -1 then
                  begin
                    allnum := False;
                    Break;
                  end;
                end;
                if allnum then
                begin
                  PageNumber := XPath('option', v.toNode).Count;
                  Break;
                end;
              end;
          finally
            Free;
          end;
    end;
  end;
end;

function GetImageURL(const DownloadThread: TDownloadThread; const AURL: String;
  const Module: TModuleContainer): Boolean;
var
  s: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.FHTTP, DownloadThread.Task.Container do
    if GET(AppendURLDelim(FillHost(Module.RootURL, AURL)) + IncStr(DownloadThread.WorkId) + '/') then
    begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do
        try
          if Module.Website = 'ReadHentaiManga' then
            s := HTMLDecode(XPathString('//img[@id="main_img"]/@src'))
          else
            s := XPathString('//*[contains(@class,"mng_rdr")]//img/@src');
          if s = '' then
            s := XPathString('//*[@id="reader"]//img[@id="picture"]/@src');
          PageLinks[DownloadThread.WorkId] := s;
        finally
          Free;
        end;
    end;
end;

procedure RegisterModule;

  function AddWebsiteModule(AWebsite, ARootURL: String): TModuleContainer;
  begin
    Result := AddModule;
    with Result do
    begin
      Website := AWebsite;
      RootURL := ARootURL;
      SortedList := True;
      OnGetDirectoryPageNumber := @GetDirectoryPageNumber;
      OnGetNameAndLink := @GetNameAndLink;
      OnGetInfo := @GetInfo;
      OnGetPageNumber := @GetPageNumber;
      OnGetImageURL := @GetImageURL;
    end;
  end;

begin
  AddWebsiteModule('MangaBoom', 'http://www.mangaboom.com');
  AddWebsiteModule('Authrone', 'http://www.authrone.com');
  AddWebsiteModule('EyeOnManga', 'http://www.eyeonmanga.com');
  AddWebsiteModule('ReadHentaiManga', 'http://readhentaimanga.com');
  AddWebsiteModule('MangaSpy', 'http://www.mangaspy.com');
  AddWebsiteModule('MangaIce', 'http://www.mangaice.com');
  AddWebsiteModule('MangaCow', 'http://mngcow.co');
  AddWebsiteModule('HentaiRead', 'http://hentairead.com');
end;

initialization
  RegisterModule;

end.

File: lua/modules/WPManga.lua
function getinfo()
  local s = ''
  mangainfo.url = MaybeFillHost(module.rooturl, url)
  if http.get(mangainfo.url) then  
    x = TXQuery.Create(http.document)
    mangainfo.coverLink = MaybeFillHost(module.rooturl, x.XPathString('//img[starts-with(@class,"cvr")]/@src'))
    mangainfo.title = x.XPathString('//*[@itemprop="itemreviewed"]')
    mangainfo.authors = x.XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Author")]/substring-after(normalize-space(.)," ")')
    mangainfo.artists = x.XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Artist")]/substring-after(normalize-space(.)," ")')
    mangainfo.status = MangaInfoStatusIfPos(x.XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Status")]/substring-after(normalize-space(.)," ")'))
    mangainfo.summary = x.XPathString('//div[@class="det"]/p[1]')
    if (module.website == 'ReadHentaiManga') then
      mangainfo.genres = x.XPathString('string-join(//*[contains(@class,"mng_det")]//*[self::p or self::li]//a,", ")')
    else
      mangainfo.genres = x.XPathString('//*[contains(@class,"mng_det")]//*[self::p or self::li][starts-with(.,"Category")]/string-join((./*[position()>1]),", ")')
    end
    if module.website == 'MangaOnlineToday' then
      mangainfo.summary = x.XPathString('//div[contains(@class,"mng_det")]/p[1]')
      x.xpathhrefall('//ul[@class="chp_lst"]/li/a', mangainfo.chapterLinks, mangainfo.chapterNames)
    else
      while true do      
        v = x.XPath('//a[@class="lst"]')
        for i = 1, v.count do
          v2 = v.get(i)
          mangainfo.chapterLinks.Add(v2.getAttribute('href'))
          s = v2.getAttribute('title')
          if s == '' then
            s = x.XPathString('*[@class="val"]', v2)
          end
          if s == '' then
            s = x.XPathString('text()[1]', v2)
          end
          mangainfo.chapterNames.Add(s)
        end
        if http.terminated then break end
        s = Trim(x.XPathString('//*[@class="pgg"]//*[./a[@class="sel"]]/following-sibling::*[./a]/a/@href'))
        if s == '' then break end
        if http.GET(MaybeFillHost(module.rooturl, s)) then
          x.ParseHTML(http.Document)
        else
          break
        end
      end
      InvertStrings(mangainfo.chapterLinks, mangainfo.chapterNames)
    end
    return no_error
  else
    return net_problem
  end
end

function getpagenumber()
  local s = ''
  local allnum = false
  http.Cookies.Values['viewer'] = '1'
  if http.GET(AppendURLDelim(MaybeFillHost(module.RootURL, url)) .. '1') then      
    -- multi page
    x = TXQuery.Create(http.Document)
    s = x.XPathString('//script[contains(.,"imglist")]/substring-after(substring-before(.,"]"),"[")')
    if s ~= '' then
      s = '[' .. s .. ']'
    else
      s = x.XPathString('//script[contains(.,"img_lst")]/substring-after(substring-before(.,"\')"),"(\'")')
      if s ~= '' then
        s = DecodeURL(s)
      end
    end
    x.ParseHTML(s)
    x.XPathStringAll('json(*)()("url")', task.PageLinks)
    if task.PageLinks.Count == 0 then
      x.XPathStringAll('json(*)()', task.PageLinks)
    end
    
    -- single page
    if task.PageLinks.Count == 0 then
      x.ParseHTML(http.Document)
      task.PageNumber = x.XPath('(//select[@class="cbo_wpm_pag"])[1]/option').Count
      if task.PageNumber == 0 then
        task.PageNumber = x.XPath('(//select[@name="page"])[1]/option').Count
      end
      if task.PageNumber == 0 then
        v = x.XPath('//select')
        for i = 1, v.count do
          allnum = true
          v2 = x.XPath('option', v.get(i))
          for i = 1, v2.count do
            if tointeger(v2.toString) == -1 then
              allnum = false
              break
            end
          end
          if allnum then
            task.PageNumber = x.XPath('option', v).Count
            break
          end
        end
      end      
    end
    return true
  else
    return false
  end   
end

function getimageurl()
  local s = ''
  if http.GET(AppendURLDelim(MaybeFillHost(module.RootURL, url)) .. IncStr(workid) .. '/') then
    x = TXQuery.Create(http.Document)        
    if module.Website == 'ReadHentaiManga' then
      s = HTMLDecode(x.XPathString('//img[@id="main_img"]/@src'))
    else
      s = x.XPathString('//*[contains(@class,"mng_rdr")]//img/@src')
    end
    if s == '' then
      s = x.XPathString('//*[@id="reader"]//img[@id="picture"]/@src')
    end
    task.PageLinks[workid] = s
    return true
  else
    return false
  end
end

function getdirurl(website)
  local result = ''
  if (website == 'MangaSpy') or (website == 'MangaIce') then
    result = 'manga_list'
  elseif (website == 'ReadHentaiManga') then
    result = 'hentai-manga-list'
  else
    result = 'manga-list'
  end
  return '/' .. result .. '/all/any/last-added/'
end

function getdirectorypagenumber()
  if http.GET(AppendURLDelim(module.RootURL) .. getdirurl(module.website)) then
    x = TXQuery.Create(http.Document)
    page = tonumber(ReplaceRegExpr('^.*\\/(\\d+)/.*$', x.XPathString('//ul[@class="pgg"]/li[last()]/a/@href'), '$1'))
    if page == nil then
      page = 1
    end
    return true
  else
    return false
  end
end

function getnameandlink()
  local w = {
    ['MangaSpy'] = true,
    ['MangaIce'] = true,
    ['MangaDeep'] = true,
    ['Manga99'] = true
  }
  if http.GET(AppendURLDelim(module.RootURL) .. getdirurl(module.website) .. IncStr(url) .. '/') then
    x = TXQuery.Create(http.Document)
    if w[module.website] then
      x.XPathHREFAll('//*[contains(@id,"content")]//*[@class="det"]/a', links, names)
    elseif module.website == 'MangaOnlineToday' then
      x.XPathHREFAll('//*[contains(@id,"content")]//div[@class="box"]/ul/li/a', links, names)
    else
      x.XPathHREFtitleAll('//*[contains(@id,"content")]//a[./img]', links, names);
    end
    return no_error
  else
    return net_problem
  end
end

function AddWebsiteModule(name, url, category)
  local m = NewModule()
  m.website = name
  m.rooturl = url
  m.category = category
  m.sortedlist = true
  m.OnGetInfo = 'getinfo'
  m.OnGetPageNumber = 'getpagenumber'
  m.OnGetImageURL = 'getimageurl'
  m.OnGetDirectoryPageNumber = 'getdirectorypagenumber'
  m.OnGetNameAndLink = 'getnameandlink'
  return m
end

function Init()
  cat = 'H-Sites'
  AddWebsiteModule('ReadHentaiManga', 'http://readhentaimanga.com', cat)
  
  cat = "Arabic-Scanlation"
  AddWebsiteModule('3asq', 'http://www.3asq.info', cat)
end
